src/storage/wal_replay.cpp

```diff
@@ -763,58 +764,96 @@ void WriteAheadLogDeserializer::ReplayRowGroupData() {
+	D_ASSERT(info->chunk->ColumnCount() == 1 && info->chunk->data[0].GetType() == LogicalType::ROW_TYPE);
 	row_t row_ids[1];
 	Vector row_identifiers(LogicalType::ROW_TYPE, data_ptr_cast(row_ids));
 
-	auto source_ids = FlatVector::GetData<row_t>(chunk.data[0]);
+	auto source_ids = FlatVector::GetData<row_t>(info->chunk->data[0]);
 	// delete the tuples from the current table
 	TableDeleteState delete_state;
-	for (idx_t i = 0; i < chunk.size(); i++) {
+	for (idx_t i = 0; i < info->chunk->size(); i++) {
 		row_ids[0] = source_ids[i];
 		state.current_table->GetStorage().Delete(delete_state, context, row_identifiers, 1);
 	}
+	return nullptr;
 }
 
-void WriteAheadLogDeserializer::ReplayUpdate() {
-	auto column_path = deserializer.ReadProperty<vector<column_t>>(101, "column_indexes");
-
-	DataChunk chunk;
-	deserializer.ReadObject(102, "chunk", [&](Deserializer &object) { chunk.Deserialize(object); });
+std::unique_ptr<UpdateInfo> WriteAheadLogDeserializer::ReplayUpdate() {
+	auto info = make_uniq<UpdateInfo>();
+	info->column_indexes = deserializer.ReadProperty<vector<column_t>>(101, "column_indexes");
+	info->chunk = make_uniq<DataChunk>();
+	deserializer.ReadObject(102, "chunk", [&](Deserializer &object) { info->chunk->Deserialize(object); });
 
 	if (DeserializeOnly()) {
-		return;
+		return info;
 	}
 	if (!state.current_table) {
 		throw InternalException("Corrupt WAL: update without table");
 	}
 
-	if (column_path[0] >= state.current_table->GetColumns().PhysicalColumnCount()) {
+	if (info->column_indexes[0] >= state.current_table->GetColumns().PhysicalColumnCount()) {
 		throw InternalException("Corrupt WAL: column index for update out of bounds");
 	}
 
 	// remove the row id vector from the chunk
-	auto row_ids = std::move(chunk.data.back());
-	chunk.data.pop_back();
+	auto row_ids = std::move(info->chunk->data.back());
+	info->chunk->data.pop_back();
 
 	// now perform the update
-	state.current_table->GetStorage().UpdateColumn(*state.current_table, context, row_ids, column_path, chunk);
+	state.current_table->GetStorage().UpdateColumn(*state.current_table, context, row_ids, info->column_indexes,
+	                                               *info->chunk);
+	return nullptr;
 }
 
-void WriteAheadLogDeserializer::ReplayCheckpoint() {
+std::unique_ptr<ParseInfo> WriteAheadLogDeserializer::ReplayCheckpoint() {
 	state.checkpoint_id = deserializer.ReadProperty<MetaBlockPointer>(101, "meta_block");
+	auto info = make_uniq<ParseInfo>(ParseInfoType::CHECKPOINT_INFO);
+	if (DeserializeOnly()) {
+		return info;
+	}
+	return nullptr;
+}
+
+WALReader::WALReader(AttachedDatabase &database, unique_ptr<BufferedFileReader> reader_p) {
+	Connection con(database.GetDatabase());
+	reader = std::move(reader_p);
+	if (reader->Finished()) {
+		return;
+	}
+	// WAL file version, so Open() has the reader positioned at the expected offset.
+	//
+	// Simplified logic from ReplayVersion(). Note that we can't call the deserializer
```

Comment:
This feels weird, and looking at the code, I'm not sure this is currently functioning correctly on `main` either ?

<https://github.com/duckdb/duckdb/pull/14785> recently increased the `WAL_VERSION_NUMBER`, which is what gets written by `WriteVersion`.

While `ReplayState` is created with `wal_version = 1` always, which makes this check here:
```c++
		if (state_p.wal_version == 1) {
			// old WAL versions do not have checksums
			return WriteAheadLogDeserializer(state_p, stream, deserialize_only);
		}
```
always trigger, making it so that the rest of the function body is never executed

@Mytherin

Reply:
So to summarize, this is actually more correct than what happens anywhere else, and we should fix that.